src/rgw/rgw_sal_sfs.h

```diff
@@ -318,7 +318,7 @@ class SFStore : public Store {
 
   bool _bucket_exists(const std::string &name) {
     const auto it = buckets.find(name);
-    return it != buckets.cend();
+    return it != buckets.cend() && it->second->get_deleted_flag() == false;
```

Comment:
Does this mean that `name` is a key in the sqlite table?
If yes, is it not possible when a bucket has been _logically_ deleted to have another bucket with the same name created later?
This sound a bit strange to me.

Reply:
Yes, you can delete a bucket and later create another bucket with the same name, that's why we needed to change to `bucket_id`.
That new bucket will have a different `bucket_id` so any object created for it will be linked to a different bucket in the metadata.
Also, when the garbage collector is done, objects pointing to a `bucket_id` of a bucket that was deleted can be safely removed permanently (if the policy allows it).

The primary key to the buckets table is now `bucket_id`. (because `name` is no longer unique)
But the way this is implemented we never query `sqlite` in order to retrieve a particular bucket. That is normally already in memory, in the `buckets` table in `types.h` via the refresh function.

The `_refresh_buckets` method in `rgw_sal_sfs.h` doesn't load the buckets that were removed, so we'll have only the alive ones in memory.

I also added a `get_bucket_by_name` method in the `sqlite` interface, but now it returns a vector.
The reason why is because you can create a bucket with the same name of a bucket that was already deleted, so it means when querying by name you could get more than one bucket. (only 1 of them should have the `deleted` flag set to `false`)

When we change this and not all the buckets are stored in memory I think we can create a `get_alive_bucket_by_name` or similar, returning just the bucket with that name and that doesn't have the `deleted` flag set to `true`.

Reply:
That's because `buckets` is the map stored in memory and the `_refresh_buckets` will discard the buckets that were deleted.
So only 1 bucket with a particular name can be located in the map. 
Let's say name is still "unique" as only 1 bucket alive has that name.

Also, when listing buckets, deleted ones are discarded (in case you call `list_buckets` before the map was refreshed)

Reply:
If we are loading only alive buckets, this check is useless:

`&& it->second->get_deleted_flag() == false`

because we already know that in memory we are keeping alive buckets

Reply:
yeah, in fact, I added it for cases in which the refresh method was not called yet.
For example, it happened to me in the test added in the PR.
I did tests with the gateway running and had no problems, but I'm not 100% sure there is a particular case in which refresh was not called yet.

It that happens, for whatever reason, you'll get a `-ERR_BUCKET_EXISTS` error, but the bucket doesn't exist logically because it was already deleted. 
So I added it for safety.

Reply:
Ok, understood.
What happens if the refresh is not called yet and a client issues a `create` bucket operation with the name of the still-to-be-evicted-from-memory bucket?
I'm just trying to think extreme scenarios here to see if the implementation is resilient to those ;)

Reply:
ok, agreed.
Personally, I don't like implementations that suggest something not true.
In this case implementation at first sight seems to say that `name` is a key for buckets, but this is not true. 
`buckets` moreover is fragile because it is a map with a key that is not an actual `key`; this could expose to potential subtle issues.
I understand that this cannot be changed easily, because all our current implementation is based on `name` as key.
I make just a proposal: what if when a bucket is deleted we also change its name prepending `_deleted_{timestamp}_{original_name}` in this way we would preserve the uniqueness for the name avoiding potential clashes.

Reply:
Do you mean to change it also in `sqlite` or only in the map?
We can do that in the map. Those entries will get removed from the map when refresh is hit, but they will not be accessible meanwhile.

One thing is how information is stored (bucket_id) and another this is how you access it (name, because users normally use the name). 
Changing that map to use (bucket_id) as the key would involve to translate from the name provided by the user to the bucket_id, which means accessing `sqlite` to query what's the bucket with that name and that is not deleted. But we have that already in memory (at least for now).

I agree it sounds weird, but we can no longer store by name... but we still access by name :/

Reply:
yeah, I understand the problem; my concerns are more on `buckets` map that has the name as key.
If you think is it worth to do I would try to ensure the uniqueness of the `name` at least in memory (using the trick of prepending something to the name).

Reply:
Ok. maybe we don't need to change the name... just removing the entry from the map would be enough.
Metadata would be already updated, so the garbage collector will be able to delete it permanently when needed and we don't need to wait for the refresh thing to remove the entry.

Name will be unique because only 1 bucket with a particular name is allowed.

Reply:
yeah, after deleting the bucket from memory that's no longer needed